source/postprocess/geoid.cc

```diff
@@ -0,0 +1,975 @@
+
+        for (unsigned int n = 2; n <= max_degree; ++n)
+          {
+            coefficients.sine_coefficients[n] = s[n] * coefficients.sine_coefficients[n] +
+                                                a[n] * M.get_coefficients().sine_coefficients[n];
+            coefficients.cosine_coefficients[n] = s[n] * coefficients.cosine_coefficients[n] +
+                                                  a[n] * M.get_coefficients().cosine_coefficients[n];
+          }
+      }
+
+      //Version of sadd which has a different factor for each degree.  As such, s and a should have sizes of (max_degree+1)
+      template <>
+      void MultipoleExpansion<3>::sadd( const std::vector<double> &s, const std::vector<double> &a, const MultipoleExpansion &M )
+      {
+        AssertThrow( coefficients.sine_coefficients.size() == M.get_coefficients().sine_coefficients.size() ,
+                     ExcInternalError() );
+
+        AssertThrow( s.size() == max_degree+1 , ExcInternalError() );
+        AssertThrow( a.size() == max_degree+1 , ExcInternalError() );
+
+        for (unsigned int l = 2, k = 0; l <= max_degree; ++l)
+          for (unsigned int m = 0; m <= l; ++m, ++k)
+            {
+              coefficients.sine_coefficients[k] = s[l] * coefficients.sine_coefficients[k] +
+                                                  a[l] * M.get_coefficients().sine_coefficients[k];
+              coefficients.cosine_coefficients[k] = s[l] * coefficients.cosine_coefficients[k] +
+                                                    a[l] * M.get_coefficients().cosine_coefficients[k];
+            }
+      }
+
+      template <int dim>
+      void
+      MultipoleExpansion<dim>::mpi_sum_coefficients (MPI_Comm mpi_communicator)
+      {
+        dealii::Utilities::MPI::sum(coefficients.sine_coefficients,mpi_communicator,coefficients.sine_coefficients);
+        dealii::Utilities::MPI::sum(coefficients.cosine_coefficients,mpi_communicator,coefficients.cosine_coefficients);
+      }
+
+
+
+      /* Implementations for doing an external multipole expansion */
+
+      template <int dim>
+      ExternalMultipoleExpansion<dim>::ExternalMultipoleExpansion(const unsigned int max_degree,
+                                                                  const double evaluation_radius)
+        : MultipoleExpansion<dim>(max_degree),
+          evaluation_radius(evaluation_radius)
+      {}
+
+      template <>
+      void
+      ExternalMultipoleExpansion<2>::add_quadrature_point (const Point<2> &position,
+                                                           const double value,
+                                                           const double weight)
+      {
+        const double r = position.norm();
+        const double theta = std::atan2(position[1],position[0]);
+
+        Assert(r/evaluation_radius -1. <= 1.e-7 ,
+               ExcMessage("Can't do an external multipole expansion with this radius") );
+
+        if ( evaluation_radius > 0.)
```

Comment:
This prevents a potential divide by zero error. I don't think it should really ever come up, since the evaluation radius has to be greater than the radii of all the quadrature points, which are all >= 0.